bgpv2: Configuration guide for BGPv2 APIs

harsimran-pabla opened the pull request:
Introducing configuration guide for managing BGP control plane. 

The first commit introduces the new documentation. The second commit is refactoring the BGP documentation structure.

harsimran-pabla committed:
bgpv2: update documentation to introduce bgpv2

Introducing BGPv2 configuration guide. It includes how to configure
BGP control plane with new APIs like
- Cilium BGP Cluster Config
- Cilium BGP Peer Config
- Cilium BGP Advertisements

Various snippets of configuration are added with this documentation.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgp-control-plane: update documentation structure

Adding bgp-control-plane directory under Documentation/network, since
there are multiple documents related to BGP control plane and it makes
it easier to navigate.

Second change is to have dedicated BGP control plane landing page which
then have references to configuration and other guides.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v1.rst:
```diff
@@ -4,56 +4,14 @@
-Agent is configured to use only IPv6 address family. The opposite is also true.
+.. warning::
+    ``CiliumBGPPeeringPolicy`` will be discontinued in future. Consider
+    using new :ref:`BGP APIs <bgp_control_plane_v2>` to configure the BGP Control Plane.
```
Nit

```suggestion
    using the new :ref:`BGP APIs <bgp_control_plane_v2>` to configure the BGP Control Plane.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane.rst:
```diff
@@ -0,0 +1,80 @@
+
+There are two ways to configure the BGP Control Plane. Using ``CiliumBGPPeeringPolicies`` resource,
+or using newer BGP resources like ``CiliumBGPClusterConfig``. Currently, both configuration options are
+supported, however ``CiliumBGPPeeringPolicy`` will be deprecated in the near future.
```
```suggestion
supported, however ``CiliumBGPPeeringPolicy`` will be deprecated in the future.
```

“near” sounds vague to me. Ideally, would you have a target Cilium version for deprecation?


Another comment: We have both `CiliumBGPPeeringPolicy` and `CiliumBGPPeeringPolicies` (plural) in this paragraph. Could we stick to just one form, to make it very clear that we're talking of the same thing?

harsimran-pabla replied:
It is tentative, we are planning its removal in 1.18, and marking it deprecated in 1.17 release.

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane.rst:
```diff
@@ -0,0 +1,80 @@
+IPv4/IPv6 single-stack and dual-stack setup are supported. Note that the BGP
+Control Plane can only advertise the route of the address family that the
+Cilium is configured to use. You cannot advertise IPv4 routes when the Cilium
+Agent is configured to use only IPv6 address family. The opposite is also true.
```
I know you just copy-pasted that part, but let's take this chance to make the note very explicit:

```suggestion
Agent is configured to use only IPv6 address family. Conversely, you cannot advertise ... when ...
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+BGP Control Plane Resources
+###########################
+
+Cilium BGP control plane is managed by set of custom resources which provide flexible way to
```
```suggestion
Cilium BGP control plane is managed by a set of custom resources which provide a flexible way to
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+Cilium BGP control plane is managed by set of custom resources which provide flexible way to
+configure BGP peers, policies, and advertisements.
+
+Following resources are used to manage BGP Control Plane
```
```suggestion
The following resources are used to manage the BGP Control Plane:
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+Following resources are used to manage BGP Control Plane
+
+* ``CiliumBGPClusterConfig``: It defines BGP instances and peer configurations which will be applied to multiple nodes.
+* ``CiliumBGPPeerConfig``: It is common set of BGP peering setting, it can be used across multiple peers.
```
```suggestion
* ``CiliumBGPPeerConfig``: A common set of BGP peering setting. It can be used across multiple peers.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+* ``CiliumBGPClusterConfig``: It defines BGP instances and peer configurations which will be applied to multiple nodes.
+* ``CiliumBGPPeerConfig``: It is common set of BGP peering setting, it can be used across multiple peers.
+* ``CiliumBGPAdvertisements``: It defines prefixes which will be injected into BGP routing table.
```
```suggestion
* ``CiliumBGPAdvertisements``: Defines prefixes that are injected into the BGP routing table.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+Following resources are used to manage BGP Control Plane
+
+* ``CiliumBGPClusterConfig``: It defines BGP instances and peer configurations which will be applied to multiple nodes.
```
```suggestion
* ``CiliumBGPClusterConfig``: Defines BGP instances and peer configurations that are applied to multiple nodes.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+* ``CiliumBGPClusterConfig``: It defines BGP instances and peer configurations which will be applied to multiple nodes.
+* ``CiliumBGPPeerConfig``: It is common set of BGP peering setting, it can be used across multiple peers.
+* ``CiliumBGPAdvertisements``: It defines prefixes which will be injected into BGP routing table.
+* ``CiliumBGPNodeConfigOverride``: It defines node specific BGP configuration which users may want to fine control.
```
```suggestion
* ``CiliumBGPNodeConfigOverride``: Defines node-specific BGP configuration to provide a finer control.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+``CiliumBGPClusterConfig`` resource is used to define BGP configuration for one or more nodes in
+the cluster based on its ``nodeSelector`` field. Each ``CiliumBGPClusterConfig`` defines one or
+more BGP instances, which are uniquely identified by ``name`` field.
```
```suggestion
more BGP instances, which are uniquely identified by their ``name`` field.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+autonomous number and peer address are defined by ``peerASN`` and ``peerAddress`` fields,
+respectively. Configuration of the peers is defined by ``peerConfigRef`` field, which is a reference
+to a peer configuration resource. ``Group`` and ``kind`` in ``peerConfigRef`` are optional and defaults to
+``cilium.io`` and ``CiliumBGPPeerConfig`` respectively.
```
```suggestion
A BGP instance can have one or more peers. Each peer is uniquely identified by its ``name`` field. The peer
autonomous number and peer address are defined by the ``peerASN`` and ``peerAddress`` fields,
respectively. The configuration of the peers is defined by the ``peerConfigRef`` field, which is a reference
to a peer configuration resource. ``Group`` and ``kind`` in ``peerConfigRef`` are optional and default to
``cilium.io`` and ``CiliumBGPPeerConfig``, respectively.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+    ``CiliumBGPPeeringPolicy`` will take precedence.
+
+
+.. code-block:: yaml
```
Please introduce this code block: “Here is an example configuration...”

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+BGP Peer Configuration
+======================
+
+``CiliumBGPPeerConfig`` resource is used to define BGP peer configuration. Multiple peers can
```
```suggestion
The ``CiliumBGPPeerConfig`` resource is used to define a BGP peer configuration. Multiple peers can
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+share the same configuration and provide reference to the common ``CiliumBGPPeerConfig``
+resource.
+
+It contains configuration options for
```
```suggestion
A ``CiliumBGPPeerConfig`` resource contains configuration options for:
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+      timers:
+      authSecretRef:
+      gracefulRestart:
+      families:
```
Please introduce this snippet as well.

Should the values in the fields be provided?

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+Transport section of ``CiliumBGPPeerConfig`` can be used to configure custom source and
+destination ports for the BGP session with the peer. ``LocalPort`` and ``PeerPort`` fields are used to
+configure source and destination ports respectively.
```
```suggestion
The transport section of ``CiliumBGPPeerConfig`` can be used to configure custom source and
destination ports for the BGP session with the peer. ``LocalPort`` and ``PeerPort`` fields are used to
configure the source and destination ports, respectively.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+configure source and destination ports respectively.
+
+By default, when BGP is operating in `active mode <https://datatracker.ietf.org/doc/html/rfc4271#section-8.2.1>`_
+( Cilium agent initiating TCP connection), destination port is 179 and source port is dynamic.
```
```suggestion
(with the Cilium agent initiating the TCP connection), destination port is 179 and source port is dynamic.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+By default, when BGP is operating in `active mode <https://datatracker.ietf.org/doc/html/rfc4271#section-8.2.1>`_
+( Cilium agent initiating TCP connection), destination port is 179 and source port is dynamic.
+
+.. code-block:: yaml
+
+    apiVersion: cilium.io/v2alpha1
+    kind: CiliumBGPPeerConfig
+    metadata:
+      name: cilium-peer
+    spec:
+      transport:
+        localPort: 179
+        peerPort: 179
```
“Here is an example...” (here and below)

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+------------
+
+``AuthSecretRef`` in ``CiliumBGPPeerConfig`` can be used to configure `RFC-2385`_ TCP MD5 password
+on the session with BGP peer which is referencing this configuration.
```
```suggestion
``AuthSecretRef`` in ``CiliumBGPPeerConfig`` can be used to configure an `RFC-2385`_ TCP MD5 password
on the session with the BGP peer which references this configuration.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+Because TCP MD5 passwords sign the header of the packet they cannot be used if
+the session will be address translated by Cilium (i.e. the Cilium Agent's pod
+IP address must be the address the BGP peer sees).
```
```suggestion
the session is address-translated by Cilium (in other words, the Cilium Agent's pod
IP address must be the address that the BGP peer sees).
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+the session will be address translated by Cilium (i.e. the Cilium Agent's pod
+IP address must be the address the BGP peer sees).
+
+If the password is incorrect, or the header is otherwise changed the TCP
```
```suggestion
If the password is incorrect, or if the header is otherwise changed, then the TCP
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+KeepaliveTimer    ``keepAliveTimeSeconds``     30
+================= ============================ ==========
+
+In datacenter networks which Kubernetes clusters are deployed, it is generally
```
```suggestion
In datacenter networks where Kubernetes clusters are deployed, it is generally
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+The Cilium BGP Control Plane can be configured to act as a graceful restart
+``Restarting Speaker``. When you enable graceful restart, the BGP session will restart
+and the "graceful restart" capability will be advertised in the BGP OPEN message.
```
```suggestion
``Restarting Speaker``. When you enable graceful restart, the BGP session restarts
and the "graceful restart" capability is advertised in the BGP OPEN message.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+By default, if no address families are specified, BGP Control Plane sends both IPv4 Unicast and IPv6 Unicast
+Multiprotocol Extensions Capability (`RFC-4760`_) to the peer.
+
+In each address family, we can control the route publication via ``advertisements`` label selector.
```
```suggestion
In each address family, you can control the route publication via the ``advertisements`` label selector.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+``CiliumBGPAdvertisement`` resource is used to define various advertisement types and attributes
+associated with them. Address family ``advertisements`` label selector in :ref:`peer configuration
+<bgp_peer_configuration_afi>` may match one or more of advertisement resources.
```
```suggestion
The ``CiliumBGPAdvertisement`` resource is used to define various advertisement types and attributes
associated with these types. Address family ``advertisements`` label selector in :ref:`peer configuration
<bgp_peer_configuration_afi>` may match one or more advertisement resources.
```

What does “Address family ``advertisements`` label selector” mean?
I find it a bit unclear, could you rephrase or detail?

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+--------------
+Users can configure BGP path attributes to the prefixes advertised by Cilium BGP
+control plane using ``attributes`` field in ``advertisements[*]``. There are two types of Path
+Attributes that can be advertised - ``Communities`` and ``LocalPreference``.
```
```suggestion
You can configure BGP path attributes for the prefixes advertised by Cilium BGP
control plane using ``attributes`` field in ``advertisements[*]``. There are two types of Path
Attributes that can be advertised: ``Communities`` and ``LocalPreference``.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+The values can be of three types:
+
+ - ``Standard``: represents a value of the "standard" 32-bit BGP Communities Attribute (`RFC-1997`_)
+   as a 4-byte decimal number or two 2-byte decimal numbers separated by a colon (e.g. ``64512:100``).
```
```suggestion
   as a 4-byte decimal number or two 2-byte decimal numbers separated by a colon (for example: ``64512:100``).
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+ - ``Standard``: represents a value of the "standard" 32-bit BGP Communities Attribute (`RFC-1997`_)
+   as a 4-byte decimal number or two 2-byte decimal numbers separated by a colon (e.g. ``64512:100``).
+ - ``WellKnown``: represents a value of the "standard" 32-bit BGP Communities Attribute (`RFC-1997`_)
+   as a well-known string alias to its numeric value. Allowed values and their mapping to the numeric values:
```
```suggestion
   as a well-known string alias to its numeric value. Allowed values and their mapping to the numeric values are displayed in the following table:
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+    =============================== ================= =================
+
+ - ``Large``: represents a value of the BGP Large Communities Attribute (`RFC-8092`_),
+   as three 4-byte decimal numbers separated by colons (e.g. ``64512:100:50``).
```
```suggestion
   as three 4-byte decimal numbers separated by colons (for example: ``64512:100:50``).
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+Advertisement Types
+-------------------
+
+Following advertisement types are supported by Cilium
```
```suggestion
The following advertisement types are supported by Cilium:
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+BGP Control Plane can advertise Pod CIDR prefixes of the nodes. This allows the BGP peers and
+connected network to reach the Pods directly without involving load balancers or NAT. There are
+two ways to advertise PodCIDRs depending on the IPAM mode setting.
```
```suggestion
The BGP Control Plane can advertise the Pod CIDR prefixes of the nodes. This allows the BGP peers and the
connected network to reach the Pods directly without involving load balancers or NAT. There are
two ways to advertise PodCIDRs depending on the IPAM mode setting.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+.. note::
+
+    Cilium BGP control plane advertises pod CIDR allocated to the node and not entire range.
```
```suggestion
    Cilium BGP control plane advertises the pod CIDR allocated to the node and not the entire range.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When :ref:`Kubernetes <k8s_hostscope>` or :ref:`ClusterPool
+<ipam_crd_cluster_pool>` IPAM is used, set advertisement type to ``PodCIDR``
```
```suggestion
When :ref:`Kubernetes <k8s_hostscope>` or :ref:`ClusterPool
<ipam_crd_cluster_pool>` IPAM is used, set advertisement type to ``PodCIDR``.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+        - advertisementType: "PodCIDR"
+
+With this configuration, BGP instance on the node advertises the
+Pod CIDR prefixes assigned to the local node.
```
```suggestion
With this configuration, the BGP instance on the node advertises the
Pod CIDR prefixes assigned to the local node.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+              pool: "blue"
+
+This advertises the PodCIDR prefixes allocated from the selected
+cilium pod IP pools. Note that the CIDR must be allocated to a ``CiliumNode`` resource.
```
```suggestion
This configuration advertises the PodCIDR prefixes allocated from the selected
Cilium pod IP pools. Note that the CIDR must be allocated to a ``CiliumNode`` resource.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+cilium pod IP pools. Note that the CIDR must be allocated to a ``CiliumNode`` resource.
+
+If you wish to announce ALL CiliumPodIPPool CIDRs within the cluster, a ``NotIn`` match
+expression with a dummy key and value can be used like:
```
```suggestion
If you wish to announce *all* CiliumPodIPPool CIDRs within the cluster, a ``NotIn`` match
expression with a dummy key and value can be used like this:
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+            - {key: somekey, operator: NotIn, values: ['never-used-value']}
+
+
+There are two special purpose selector fields that match CiliumPodIPPools based on ``name`` and/or
```
```suggestion
There are two special-purpose selector fields that match CiliumPodIPPools based on ``name`` and/or
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+~~~~~~~~~~~~~~~~
+
+When using other IPAM types, the BGP Control Plane does not support advertising
+PodCIDRs and specifying ``advertisementType: "PodCIDR"`` doesn't take any
```
```suggestion
PodCIDRs and specifying ``advertisementType: "PodCIDR"`` doesn't have any
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+or ``.spec.externalIPs``.
+
+The BGP control plane can advertise the virtual IP address of the Service to BGP peers.
+This allows users to directly access the Service from outside the cluster.
```
```suggestion
This allows you to directly access the Service from outside the cluster.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+This allows users to directly access the Service from outside the cluster.
+
+.. note::
+    Cilium BGP Control Plane advertised exact routes for the VIPs ( /32 or /128 prefixes ).
```
```suggestion
    Cilium BGP Control Plane advertises exact routes for the VIPs (/32 or /128 prefixes).
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+When your upstream router supports Equal Cost Multi Path(ECMP), you can use
+this feature to load balance traffic to the Service across multiple nodes by
+advertising the same virtual IPs from multiple nodes.
```
```suggestion
When your upstream router supports Equal Cost Multi Path (ECMP), you can use
this feature to load-balance traffic to the Service across multiple nodes by
advertising the same virtual IPs from multiple nodes.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+If you wish to use this together with ``kubeProxyReplacement`` feature  (see :ref:`kubeproxy-free` docs),
+please make sure the ExternalIP support is enabled.
+
+If you only wish to advertise the ``.spec.externalIPs`` of Service, you can specify the
```
```suggestion
If you only wish to advertise the ``.spec.externalIPs`` of a Service, you can specify the
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+See :ref:`External Access To ClusterIP Services <external_access_to_clusterip_services>` section
+for how to enable it.
+
+If you only wish to advertise the ``.spec.clusterIP`` and ``.spec.clusterIPs`` of Service,
```
```suggestion
If you only wish to advertise the ``.spec.clusterIP`` and ``.spec.clusterIPs`` of a Service,
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+This advertises the ingress IPs of all Services matching the ``.selector``.
+
+If you wish to announce ALL services within the cluster, a ``NotIn`` match expression
+with a dummy key and value can be used like:
```
```suggestion
If you wish to announce *all* services within the cluster, a ``NotIn`` match expression
with a dummy key and value can be used like this:
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+            matchExpressions:
+             - {key: somekey, operator: NotIn, values: ['never-used-value']}
+
+There are a few special purpose selector fields which don't match on labels but
```
```suggestion
There are a few special purpose selector fields that don't match on labels but
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+Cilium supports the `loadBalancerClass
+<https://kubernetes.io/docs/concepts/services-networking/service/#load-balancer-class>`__.
+When the load balancer class is set to ``io.cilium/bgp-control-plane`` or unspecified,
+Cilium will announce the ingress IPs of the Service. Otherwise, Cilium will not announce
```
```suggestion
Cilium announces the ingress IPs of the Service. Otherwise, Cilium does not announce
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+==========================
+
+``CiliumBGPNodeConfigOverride`` can be used to override some of the auto generated configuration
+on a per node basis.
```
```suggestion
``CiliumBGPNodeConfigOverride`` can be used to override some of the auto-generated configuration
on a per-node basis.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+.. note::
+    Name of ``CiliumBGPNodeConfigOverride`` resource must match the name of the node for which configuration is intended.
+    Similarly, names of the BGP instance and peers must match with what is defined under ``CiliumBGPClusterConfig``.
```
```suggestion
    The name of ``CiliumBGPNodeConfigOverride`` resource must match the name of the node for which the configuration is intended.
    Similarly, the names of the BGP instance and peers must match with what is defined under ``CiliumBGPClusterConfig``.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+the IPv4 address assigned to the node as the BGP Router ID because Router ID is 32bit long, and
+we can rely on the uniqueness of the IPv4 address to make Router ID unique which is not the case
+for IPv6. Thus, when running in an IPv6 single-stack, or when the auto assignment of the Router ID
+is not desired, the administrator needs to manually define it.
```
```suggestion
When Cilium runs on an IPv4 single-stack or a dual-stack, the BGP Control Plane can use
the IPv4 address assigned to the node as the BGP Router ID because the Router ID is 32 bit-long, and
we can rely on the uniqueness of the IPv4 address to make the Router ID unique, which is not the case
for IPv6. Thus, when running in an IPv6 single-stack, or when the auto assignment of the Router ID
is not desired, the administrator needs to manually define it.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+for IPv6. Thus, when running in an IPv6 single-stack, or when the auto assignment of the Router ID
+is not desired, the administrator needs to manually define it.
+
+In order to configure custom router-id, user can set ``routerID`` field in an IPv4 address format.
```
```suggestion
In order to configure custom Router ID, the user can set ``routerID`` field in an IPv4 address format.
```

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+
+Source interface and address used by the BGP Control Plane in order to setup peering with the neighbor is based on route
+lookup of peer address defined in ``CiliumBGPClusterConfig``. There may be use cases where multiple links
+are present on the node and user wants tighter control over which link BGP peering should be setup.
```
```suggestion
The source interface and the address used by the BGP Control Plane in order to setup peering with the neighbor are based on a route
lookup of the peer address defined in ``CiliumBGPClusterConfig``. There may be cases where multiple links
are present on the node and the user wants tighter control over which link BGP peering should be setup.
```

(Please wrap on ~80 characters)

qmonnet commented on Documentation/network/bgp-control-plane/bgp-control-plane-v2.rst:
```diff
@@ -0,0 +1,765 @@
+are present on the node and user wants tighter control over which link BGP peering should be setup.
+
+To configure source address, ``peers[*].localAddress`` field can be set. It should be an address configured on one of
+the links on the node.
```
```suggestion
To configure the source address, the ``peers[*].localAddress`` field can be set. It should be an address configured on one of
the links on the node.
```

qmonnet reviewed (changes requested):
Nice work, thanks!

As for the split, I'd find more logical to have the file renaming/moving first, and then to add the new doc, rather than adding it to rename it straight after.

Please find some minor comments and suggestions inline below.

networkop commented on Documentation/network/bgp-control-plane/bgp-control-plane.rst:
```diff
@@ -0,0 +1,80 @@
+    :glob:
+
+    bgp-control-plane-v1
+    bgp-control-plane-v2
```
should we move v2 to the top here?

qmonnet reviewed (approved):
Thanks a lot!

harsimran-pabla committed:
bgpv2/docs: adding BGP resource diagram

Adding bgpv2.png file which shows relationship between various BGP
configuration resources.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

networkop reviewed (approved):
lgtm, thanks!

rastislavs reviewed (approved):
Awesome work, thanks!
